### ISSUE TITLE
Migrate Fabric to Needed Versions for Kube 1.6

### ISSUE THREAD

**Issue by mikeln** [samsung-cnct/k2#435]
This has the setup we should use:
https://github.com/projectcalico/canal/tree/master/k8s-install/1.6

Research has shown we just pulled in (and fixed) this file originally:
https://github.com/projectcalico/canal/blob/master/k8s-install/canal_etcd_tls.yaml

Now, the current recommendations/setup for v1.6 are different.   Migrate our fabric to the new setup.

(This should also fix this issue with DNS startup: Kubelet starts pods before CNI is configured)

**Comment by mikeln** [comment]
```
Requirements:

Make sure your k8s cluster is configured to provide serviceaccount tokens to pods.
Make sure your kubelets have been started with --network-plugin=cni and have --cni-conf-dir and --cni-bin-dir properly set
Make sure your controller manager has been started with --cluster-cidr=10.244.0.0/16 and --allocate-node-cidrs=true.
```

**Comment by mikeln** [comment]
Set those values to our config values.

need to test rbac and non-rbac running of 1.6

**Comment by mikeln** [comment]
DNS having issues.  rbac is part of the problem...not sure the CIDRs are correct.

**Comment by mikeln** [comment]
ok, this all seems to be working now.

**Comment by mikeln** [comment]
Ok, I think we need to set up this configuration like I did KubeDNS.   The configuration in kubeDNS was very different for each kubernetes release.  It was impossible to maintain all the same config vars, so role defaults was set up to deal with the specific version defaults.  The user can still override them if wanted.

A similar setup should be used here for fabric pre-1.6 vs post-1.6.

**Comment by mikeln** [comment]
e.g. given:
```
  fabricConfigs:
    - &defaultCanalFabric
      name: defaultCanalFabric
      kind: fabric
      type: canal
      options:
        containers:

 - &defaultCanalFabric16
      name: defaultCanalFabric16
      kind: fabric
      type: canal
      options:
        containers:

deployment:
  clusters:
    fabricConfig: *defaultCanalFabric
```
perhaps:
```
deployment:
  clusters:
    fabricConfig:
       default: *defaultCanalFabric
       versions:
         - v1.6: *defaultCanalFabric16
         - v1.7: *defaultCanalFabric17
```
rules:  if a `versions[ <kube minor version> ] `exists...use it.  else use `default`

Issues: 

1. version value semantics - always vX.Y would be required.  Should be ok since semver is the usage.
2. version as a key - are periods allowed in keys?  - v1.6 ?

**Comment by davidewatson** [comment]
Periods are allowed in keys (YAML's pretty flexible in this regard; you can quote keys too).

I am concerned about adding implicit relationships between stanzas (e.g. value points to key/stanza). One reason is JSON schema draft 4 does not provide a mechanism to validate these relationships (though draft 5 might with $data). More importantly, they make it difficult to understand the semantics of individual value changes and this was one of the major motivations for k2recon.

providerConfig is a good example of where this has become a problem in the past. Since there are provider specific aspects to both the cluster and node configuration, we created two different stanzas: clusters[0].providerConfig and cluster[0].nodePools[0].nodeConfig.providerConfig.

Given the k8s specific nature of the fabric configuration, I wonder if it might not be better to reference it from kubeConfig? Grouping dependent values together should make it easier to see the relationship between them.

**Comment by mikeln** [comment]
Ok, I agree with that.  It is all common, and not cluster specific, so yes, in the config area would make more sense.  I also like the current anchor setup, but don't want to make it so a user would have to duplicate the whole setup for all versions to change one value...so how about this:
```
fabricConfigs:
    - &defaultCanalFabric
      name: defaultCanalFabric
      kind: fabric
      type: canal
      options:
        containers:
        <more stuff>
 - &defaultCanalFabric16
      name: defaultCanalFabric16
      kind: fabric
      type: canal
      options:
        containers:
        <more stuff>
 - &kubeVersionedFabric
     name: kubeVersionedFabric
     kind: versionedFabric
     kubeVersion:
        default: *defaultCanalFabric
        versions:
          - v1.6: *defaultCanalFabric16
          - v1.7: *defaultCanalFabric17
deployment:
  clusters:
    fabricConfig: * kubeVersionedFabric
```

**Comment by mikeln** [comment]
usage goal: task to set var for template use:
* check `fabricConfig.kind`
** when `== versionedFabric`
*** when `kubeVersion.versions[ <kubeVer >] exists`
**** set var `fabricConfigVal` <== `kubeVersion.versions[ <kubeVer >]`
*** else set var `fabricConfigVal` <== `kubeVersion.default`
** else set var `fabricConfigVal` <== `fabricConfig`

template uses `fabricConfigVal`

**Comment by coffeepac** [comment]
I think making changes or selections based on a change in kubernetes version is something that we should be doing.  @davidewatson I understand it is obscured that the running config is changing when moving from 1.5 -> 1.6 as K2 will be tearing down/rebuilding the network fabric but that's something that K2 implicitly says it will do when it performs the upgrade.  Or is there something else here I'm missing?

@mikeln I like the second option better.  It keeps the cluster definition cleaner.

**Comment by davidewatson** [comment]
The concern is not primarily whether the user knows what is changing, though that may be important. My worry is that the user will not understand how the various stanzas fit together at a glance. For example, given the scheme above, suppose the user changes kubeConfig.version. How will they learn that they need to update the versionedFabricConfig stanza too?

Another example of the general problem can be seen in dnsConfig.ip_address property. Without reading the documentation, it might be difficult to discover that "This should be the same as the IP set in the deployment.clusters[x].dns in the main configuration file".

I do like the second option though. The name kubeVersionedFabric is helpful. A third option would be:

```
fabricConfigs:
    - &defaultCanalFabric16
      name: defaultCanalFabric16
      kind: fabric
      type: canal
      options:
        containers:
...
  kubeConfigs:
    - &defaultKube
      name: defaultKube
      kind: kubernetes
      version: v1.6.0
      fabricConfig: *defaultCanalFabric16
      hyperkubeLocation: gcr.io/google_containers/hyperkube
...
deployment:
  clusters:
    - name: myCluster
      nodePools:
       - name: master
        kubeConfig: *defaultKube
```

**Comment by coffeepac** [comment]
Ahh, gotcha.  I was already operating as if we had accepted @mikeln 's suggestion that we make which fabric gets used transparent to the user.  

I don't like the solution you are proposing for two reasons:
- the network fabric is independent of kubernetes versioning.  We are using CNI plugins which are not kubernetes specific
- not having the default fabric config point to a selector implies that each kubernetes version will require its own entry in the fabricConfigs list.  this is hopefully not true.

**Comment by davidewatson** [comment]
You're right that the "network:" stanza of fabricConfig is independent and would be duplicated unless we break it out, but the "containers:" stanza is dependent on kubernetes version. So we effectively have a separate containers stanza for each k8s version in both cases. The only difference is whether or not that link is explicit (using YAML anchors and aliases) or implicit (using a selector).

I see the appeal of creating a selector. I am just worried about the notation being a one-off solution. If we use a selector for the fabricCOnfig/kubeConfig dependency, then ideally the syntax would be generic enough to reuse for nodeConfig/providerConfig dependency. Both nodeConfig and fabricConfig are currently a mix of dependent and independent values.

**Comment by coffeepac** [comment]
closed by PR #435
